File: 0138-copy-list-with-random-pointer/0138-copy-list-with-random-pointer.java
/*
// Definition for a Node.
class Node {
    int val;
    Node next;
    Node random;

    public Node(int val) {
        this.val = val;
        this.next = null;
        this.random = null;
    }
}
*/

class Solution {
    public Node copyRandomList(Node head) {
        Map<Node, Node> copyMap = new LinkedHashMap<>();
        Node headNode = head;
        Node headNodeCopy = head;
        Node deepCopy = new Node(0);
        Node deepCopyHeadNode = deepCopy;

        while (headNode != null) {
            copyMap.put(headNode, new Node(headNode.val));
            headNode = headNode.next;
        }

        for (Map.Entry<Node, Node> entry: copyMap.entrySet()) {
            Node newNode = entry.getValue();
            newNode.next = copyMap.get(headNodeCopy.next);
            newNode.random = copyMap.get(headNodeCopy.random);
            deepCopyHeadNode.next = newNode;
            headNodeCopy = headNodeCopy.next;
            deepCopyHeadNode = deepCopyHeadNode.next;
        }

        return deepCopy.next;
    }
}

File: 0138-copy-list-with-random-pointer/0138-copy-list-with-random-pointer.py
"""
# Definition for a Node.
class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random
"""

class Solution:
    def copyRandomList(self, head: 'Optional[Node]') -> 'Optional[Node]':
        headNode = head
        dummy = dummyNode = res = Node(0)
        hashMap = {}

        while headNode:
            dummy.next = Node(headNode.val)
            hashMap[headNode] = dummy.next
            dummy = dummy.next
            headNode = headNode.next
        
        dummyNode = dummyNode.next
        while head:
            if head.random:
                dummyNode.random = hashMap[head.random]
            dummyNode = dummyNode.next
            head = head.next
        
        return res.next
